File: algospot/fossile.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <algorithm>

using namespace std;

struct point
{
	double x, y;
};

typedef pair<point, point> pp;
typedef vector<pp> vpp;

int main()
{
#ifdef _DEBUG
	freopen("fossile.in", "r", stdin);
#endif
	
	int C, N, M;
	cin >> C;
	while (C--)
	{
		cin >> N >> M;

		vector<point> hull[2];
		hull[0].resize(N);
		hull[1].resize(M);
		for (auto& h : hull[0])
			cin >> h.x >> h.y;
		for (auto& h : hull[1])
			cin >> h.x >> h.y;

		vpp upper, lower;

		auto decompose = [&upper, &lower](const vector<point>& hull) {
			int n = hull.size();
			// divide into upper, lower hull 
			for (int i = 0; i < n; i++) {
				auto& a = hull[i];
				auto& b = hull[(i+1)%n];
				if (a.x > b.x)
					upper.push_back({ a, b });
				else if (a.x < b.x)
					lower.push_back({ a, b });
			}
		};
		decompose(hull[0]);
		decompose(hull[1]);

		auto inside = [&](const point& p, const point& q, double x) {
			return (p.x <= x && x <= q.x) || (q.x <= x && x <= p.x);
		};

		auto at = [](const point& p, const point& q, double x) {
			double dy = q.y - p.y, dx = q.x - p.x;
			return p.y + dy * (x - p.x) / dx;
		};

		auto vertical = [&](double x)
		{
			double minY = 1e20;
			for (auto& u : upper)
				if (inside(u.first, u.second, x))
					minY = min(minY, at(u.first, u.second, x));

			int ln = lower.size();
			double maxY = -1e20;
			for (auto& l : lower)
			{
				if (inside(l.first, l.second, x))
					maxY = max(maxY, at(l.first, l.second, x));
			}
			return minY - maxY;
		};

		// solve
		auto minX = [](const vector<point>& p)
		{
			double v = 1e20;
			for (auto i = 0U; i < p.size(); i++)
				v = min(v, p[i].x);
			return v;
		};

		auto maxX = [](const vector<point>& p)
		{
			double v = -1e20;
			for (auto i = 0U; i < p.size(); i++)
				v = max(v, p[i].x);
			return v;
		};

		auto lo = max(minX(hull[0]), minX(hull[1]));
		auto hi = min(maxX(hull[0]), maxX(hull[1]));

		if (lo > hi)
		{
			cout << 0 << endl;
			continue;
		}

		for (int iter = 0; iter < 100; iter++)
		{
			auto a = (lo*2.0 + hi)/3.0;
			auto b = (lo + hi*2.0)/3.0;
			if (vertical(a) < vertical(b))
				lo = a;
			else
				hi = b;
		}

		cout << fixed << setprecision(10);
		cout << max(0.0, vertical(lo)) << endl;
	}
	return 0;
}

File: algospot/fossile.py
def read():
    k = map(float, raw_input().split())
    return list(zip(*[iter(k)]*2))

def low(hull):
    v = -1e20
    for h in hull:
         v = max(v, min(x for x, y in h))
    return v

def high(hull):
    v = 1e20
    for h in hull:
         v = min(v, max(x for x, y in h))
    return v

def between(a, b, x):
    x1, y1 = a
    x2, y2 = b
    return x1 <= x <= x2 or x2 <= x <= x1

def at(a, b, x):
    x1, y1 = a
    x2, y2 = b
    return y1 + (x-x1)*(y2-y1)/(x2-x1)

def vertical(x):
    minY = min(at(seg[0], seg[1], x) for seg in upper if between(seg[0], seg[1], x)) 
    maxY = max(at(seg[0], seg[1], x) for seg in lower if between(seg[0], seg[1], x)) 
    return minY - maxY

for C in xrange(input()):
    N, M = map(int, raw_input().split())
    hull = (read(), read())

    lower, upper = [], []
    for h in hull:
        for a, b in zip(h, h[1:]+h[:1]):
            if a[0] < b[0]:
                lower.append((a, b))
            elif a[0] > b[0]:
                upper.append((a, b))
    lo, hi = low(hull), high(hull)
    ans = 0
    if hi > lo:
        for i in range(100):
            a = (lo*2 + hi) / 3
            b = (lo + hi*2) / 3
            if vertical(a) < vertical(b):
                lo = a
            else:
                hi = b
        ans = max(0, vertical(lo))
    print(ans)
